application/controllers/admin/tokens.php

```diff
@@ -2148,7 +2148,47 @@ public function import($iSurveyId)
                             }
                         }
 
-                        if (!$bDuplicateFound && !$bInvalidEmail && !$bInvalidToken) {
+                        // Dispatch beforeTokenImport event
+                        $params = array(
+                            'csvcharset' => $sUploadCharset,
+                            'filterduplicatetoken' => $bFilterDuplicateToken,
+                            'filterblankemail' => $bFilterBlankEmail,
+                            'allowinvalidemail' => $bAllowInvalidEmail,
+                            'filterduplicatefields' => $aFilterDuplicateFields,
+                            'separator' => $sSeparator,
+                            'showwarningtoken' => Yii::app()->request->getPost('showwarningtoken'),
+                        );
+                        $event = new PluginEvent('beforeTokenImport');
+                        $event->set('importType', 'CSV');
+                        $event->set('surveyId', $iSurveyId);
+                        $event->set('params', $params);
+                        $event->set('recordCount', $iRecordCount);
+                        $event->set('token', $aWriteArray);
+                        $event->set('importDone', false);
+                        $event->set('importValid', true);                        
+                        App()->getPluginManager()->dispatchEvent($event);
+                        $bPluginReportedError = !$event->get('importValid');
+                        $bImportDone = $event->get('importDone');
```

Comment:
Can we split out this code section into a new function (in same file)? I don't like it when functions grow indefinitely. :|

Reply:
Something like wrapping the whole plugin call in a function?

Reply:
If possible, yes, would be great.

Reply:
Done!